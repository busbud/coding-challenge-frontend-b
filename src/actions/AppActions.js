import AppDispatcher from './AppDispatcher';
import async from 'async';
const _ = require('lodash');

class AppActions {
  fetchAll(from, to, when) { 
    this.dispatchIsLoading(true);
    // Cannot use 29 of july because date already past, so i am using a dynamic date choose by the user
    when = when.toISOString().slice(0,10);
    this.from = from;
    this.to = to;
    this.when = when;
    this.endpoint = `https://napi.busbud.com/x-departures/${from}/${to}/${when}`;

    fetch(this.endpoint, {
      headers: {
        'X-Busbud-token': 'PARTNER_JSWsVZQcS_KzxNRzGtIt1A',
        'Accept': 'application/vnd.busbud+json; version=2; profile=https://schema.busbud.com/v2/'
      }
    })
    .then((res) => res.json())
    .then((myJson) => { 
      const { cities, locations, operators, departures } = myJson;
      this.results = { cities, locations, operators, departures, travelDate: when };

      this.dispatchResults(this.results);

      if(!myJson.complete){
        this.complete = myJson.complete;
        // Store index in order to loop if call response is not complete
        this.index = myJson.departures.length;
        // Call loop to continue to fetch results
        this.fetchPartial();
      }else{
        this.dispatchIsLoading(false);
      }
    })
    .catch((err) => {
      console.error('Fetch API: An error occured');
      console.error(err);
    });
  }
  fetchPartial(){
    // Loop as long as response is not complete
    async.until(
       () => { return this.complete; },
       (cb) => {
        // Loop every x seconds (1 seconds for this test) to not overload the server
        setTimeout( () =>{
          const partialEndpoint = `${this.endpoint}/poll/?index=${this.index}`;
          fetch(partialEndpoint, {
            headers: {
              'X-Busbud-token': 'PARTNER_JSWsVZQcS_KzxNRzGtIt1A',
              'Accept': 'application/vnd.busbud+json; version=2; profile=https://schema.busbud.com/v2/'
            }
          })
          .then((res) => res.json())
          .then((myJson) => { console.log(myJson);
            const { operators, departures } = myJson;
            const { cities, locations, travelDate } = this.results;

            const refreshedOperators = _.unionBy( this.results.operators, operators, 'id');
            const refreshedDepartures = [...this.results.departures, ...departures];

            this.results = { 
              cities, 
              locations, 
              operators: refreshedOperators, 
              departures: refreshedDepartures,
              travelDate
            };

            this.complete = myJson.complete;

            if(!myJson.complete){
              // Update index for next partial call
              this.index += myJson.departures.length; 
              // Continue to loop
              cb();
            }else{
              this.dispatchIsLoading(false);
            }
            // Return results with partial results added
            this.dispatchResults(this.results);
          })
          .catch((err) => {
            console.error('Fetch API (Partial): An error occured');
            console.error(err);
          });
        }, 1000);
      },function (err) {
        console.error('ASYNC Fetch api partial: An error occured');
        console.error(err);
      }
    );
  }

  dispatchResults(data) {
    AppDispatcher.dispatch({
      actionType: 'POPULATE_RESULTS',
      results: data,
    });
    return null;
  }
  dispatchIsLoading(data) {
    AppDispatcher.dispatch({
      actionType: 'TOGGLE_LOADER',
      isLoading: data,
    });
    return null;
  }
}

export default new AppActions();
