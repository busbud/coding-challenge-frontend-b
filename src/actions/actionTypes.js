export const SET_FETCHING = 'SET_FETCHING';
export const UNSET_FETCHING = 'UNSET_FETCHING';

export const GET_DEPARTURES_SUCCESS = 'GET_DEPARTURES_SUCCESS';
export const GET_DEPARTURES_FAILURE = 'GET_DEPARTURES_FAILURE';
