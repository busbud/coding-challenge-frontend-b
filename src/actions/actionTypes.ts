// NOTE: linting disabled to allow for clear constants to be defined
/*  eslint-disable @typescript-eslint/naming-convention */
export const actionTypes = {
  DEPARTURE_FETCH_ERROR: "DEPARTURE_FETCH_ERROR",
  DEPARTURE_FETCH_START: "DEPARTURE_FETCH_START",
  DEPARTURE_FETCH_SUCCESS: "DEPARTURE_FETCH_SUCCESS",
  DEPARTURE_POLL: "DEPARTURE_POLL",
};
