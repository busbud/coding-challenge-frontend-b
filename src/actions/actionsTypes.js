export const POPULATE_DEPARTURES_SUCCESS = 'POPULATE_DEPARTURES_SUCCESS'
