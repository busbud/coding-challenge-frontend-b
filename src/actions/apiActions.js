import {
  API,
  API_START,
  API_END,
  SET_DEPARTURES_LOCATIONS,
  GET_DEPARTURES_LOCATIONS
} from '../constants/actionTypes';

export const apiStart = payload => ({
  type: API_START,
  payload
});

export const apiEnd = payload => ({
  type: API_END,
  payload
});

const setDeparturesAndLocations = payload => {
  return {
    type: SET_DEPARTURES_LOCATIONS,
    payload
  };
};

const failureHandler = error => {
  //TODO: error handling
  console.log(error);
  return {
    type: ''
  };
};

export const getDeparturesAndLocations = () => {
  return {
    type: API,
    payload: {
      label: GET_DEPARTURES_LOCATIONS,
      onSuccess: setDeparturesAndLocations,
      onFailure: failureHandler
    }
  };
};
