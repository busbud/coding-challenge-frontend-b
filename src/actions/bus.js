import api from './api/busbud.js';

const ROUTES_FETCH_BEGIN = 'ROUTES_FETCH_BEGIN';
const ROUTES_FETCH_SUCCESS = 'ROUTES_FETCH_SUCCESS';
const ROUTES_FETCH_FAILURE = 'ROUTES_FETCH_FAILURE';
export const actions = { ROUTES_FETCH_BEGIN, ROUTES_FETCH_SUCCESS, ROUTES_FETCH_FAILURE };

const fetchRoutesBegin = () => ({ type: actions.ROUTES_FETCH_BEGIN });
const fetchRoutesSuccess = (routes, isComplete) => ({ type: actions.ROUTES_FETCH_SUCCESS, routes, isComplete });
const fetchRoutesFailure = (error) => ({ type: actions.ROUTES_FETCH_FAILURE, error });

export function getRoutes(origin, destination, outbound_date) {
  return async (dispatch) => {
    dispatch(fetchRoutesBegin());
    try {
      const response = await api.getRoutes(origin, destination, outbound_date);
      const { list, complete } = response.data;
      dispatch(fetchRoutesSuccess(list, complete));
    } catch (error) {
      dispatch(fetchRoutesFailure(error));
    }
  }
}
