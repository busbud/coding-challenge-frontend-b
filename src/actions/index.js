import fetch from 'isomorphic-fetch';
import qs from 'qs';
import { isEqual, isUndefined } from 'lodash';

/**
 * toggleLang action
 **/
export const toggleLang = (lang) => {
    return {
        type: 'TOGGLE_LANG',
        lang: lang
    };
};


/**
 * toggleCurrency action
 **/
export const toggleCurrency = (currency) => {
    return {
        type: 'TOGGLE_CURRENCY',
        currency: currency
    };
};


/**
 * changeSort action
 **/
export const sortTypes = ['departureDate', 'price', 'company'];
export const changeSort = (sort) => {
    return {
        type: 'CHANGE_SORT',
        sort: sort
    };
};





/**
 *  Api actions
 **/


function fetchDepartures(params, queryParams) {
    return {
        type: 'FETCH_API_REQUEST',
        params: params,
        queryParams: queryParams
    };
}

function receiveDepartures(params, queryParams, json) { 
    return {
        type: 'FETCH_API_SUCCESS',
        params: params,
        queryParams: queryParams,
        json: json
    };
}

function failFetchingDepartures(params, queryParams, error) {
    return {
        type: 'FETCH_API_FAILURE',
        params: params,
        queryParams: queryParams,
        error: error
    };
}

let refresher;
export function createRefresher(ttl, dispatch) {
    //wait the ttl then redispatch (to keep list up-to-date)
    if (!refresher && ttl && ttl > 0) {

        const diffInMs = Math.ceil(ttl)*1000;
        refresher = setTimeout(() => {
            dispatch(fetchApiIfNeeded());
            //reset the timer
            clearTimeout(refresher);
            refresher = null;
        }, diffInMs + 1000);//Add 1 second to ensure the new data will def. be available
    }
}

/**
 *  fetchApi() : do the actual api request
 **/

function fetchApi(params, queryParams) {

    let isPoll = queryParams.hasOwnProperty('index');

    return dispatch => {
        dispatch(fetchDepartures(params, queryParams));
        return fetch(
                `https://napi.busbud.com/x-departures/${params.NY}/${params.MTL}/${params.date}${isPoll ? '/poll' : ''}?${qs.stringify(queryParams)}`,
                {
                    headers: {
                        'Accept': 'application/vnd.busbud+json; version=2; profile=https://schema.busbud.com/v2/'
                    }
                }
            )
        .then((response) => response.json())
        .then((json) => {

            //calculate the expiration date from the TTL (assuming TTL is in seconds)
            if (json && json.ttl) {

                let diffInMs = Math.ceil(json.ttl)*1000;
                json.expireDate = new Date(diffInMs + (+new Date()));

                //ceate a refresher to make sure we always have up-to-date data
                createRefresher(json.ttl, dispatch);
            }

            //wait a little bit before redispatching: 
            //for some reason, fetch() stop querying the server after ~10 consecutive tries.
            //since the poll takes sometime more time to complete the list of departures,
            //we're applying a light delay between requests:
            setTimeout(() => {
                dispatch(receiveDepartures(params, queryParams, json));

                //if the request is not complete: redispatch with polling status (index)
                if (!json.complete) {
                    let newQueryParams = Object.assign({}, queryParams, {
                        index: json.departures.length + (queryParams.index || 0)
                    });
                    
                    dispatch(fetchApiIfNeeded(params, newQueryParams));
                }
            }, json.complete ? 1 : 1000);//I know this is bad, but I can't do any other way (poor documentation of the fetch() wrapper...)
        })
        .catch(error => failFetchingDepartures(params, queryParams, error));
    };
}

function shouldFetchApi(state, params, queryParams) {

    //determine if the data is expired by comparing the previously stored property expireDate (Date) to now.
    let isExpired = !(state.api && state.api.data && state.api.data.expireDate) || new Date(state.api.data.expireDate) < new Date();

    //fetch only if the queries have changed or if the TTL expired for the current query
    if (isEqual(state.api.lang, queryParams.lang) ||
        !isEqual(state.api.currency, queryParams.currency) ||
        !isUndefined(queryParams.index) ||
        (isExpired && !state.api.isFetching)) {

        return true;
    }
    return false;
}

/**
 *  getQueryParams(): get the query params object, based on the proper parameters
 *  index must be set only if the 'complete' property is not true
 **/
export function getQueryParams(lang = 'EN', currency = 'CAD', index = null) {
    let queryParams = {
        adult : 1,
        child : 0,
        senior : 0,
        lang : lang,
        currency : currency
    };

    if (typeof index === 'number') {
        queryParams.index = index;
    }

    return queryParams;
}

export function fetchApiIfNeeded(params = {
    NY: 'dr5reg', //New York geohash
    MTL: 'f25dvk', //Montreal geohash
    date: '2016-08-29' //the 29th of July 2016
}, queryParams = getQueryParams()) {

    return (dispatch, getState) => {
        if (shouldFetchApi(getState(), params, queryParams)) {
            return dispatch(fetchApi(params, queryParams));
        }
    };
}
