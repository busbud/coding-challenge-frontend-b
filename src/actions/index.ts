import IDeparture from "../types/departure";
import IOperator from "../types/operator";
import ILocation from "../types/location";
import ICity from "../types/city";

interface UpdateLanguageAction {
  type: "UPDATE_LANGUAGE";
  language: string;
}
interface loadDeparturesAction {
  type: "LOAD_DEPARTURES";
  departures: Array<IDeparture>;
  locations: Array<ILocation>;
  operators: Array<IOperator>;
  cities: Array<ICity>;
  polling: boolean;
}
export function updateLanguage(language: string): UpdateLanguageAction {
  return { type: "UPDATE_LANGUAGE", language };
}

export function loadDepartures(
  departures: Array<IDeparture>,
  locations: Array<ILocation>,
  operators: Array<IOperator>,
  cities: Array<ICity>,
  polling: boolean
): loadDeparturesAction {
  return {
    type: "LOAD_DEPARTURES",
    departures,
    locations,
    operators,
    cities,
    polling
  };
}

function startSearching() {
  return {
    type: "START_SEARCHING"
  };
}

function doneSearching() {
  return {
    type: "DONE_SEARCHING"
  };
}

export function fetchDepartures(
  origin: string,
  destination: string,
  outboundDate: string,
  polling: boolean,
  index?: number
) {
  return function(dispatch: any) {
    const token = process.env.REACT_APP_BUSBUD_TOKEN;
    if (!token) {
      console.log("No BUSBUD_TOKEN found");
    } else {
      if (!polling) {
        dispatch(startSearching());
      }
      return fetch(
        `https://napi.busbud.com/x-departures/${origin}/${destination}/${outboundDate}${
          polling ? `/poll?index=${index}` : ""
        }`,
        {
          method: "GET",
          headers: {
            Accept:
              "application/vnd.busbud+json; version=2; profile=https://schema.busbud.com/v2/",
            "X-Busbud-Token": token
          }
        }
      )
        .then(
          response => response.json(),
          error => window.console.log("error: ", error)
        )
        .then(json => {
          if (!json.errors) {
            if (json.departures && json.departures.length > 0) {
              dispatch(
                loadDepartures(
                  json.departures,
                  json.locations,
                  json.operators,
                  json.cities,
                  polling
                )
              );
            }
            if (json.complete) {
              dispatch(doneSearching());
            } else {
              // We only want to keep polling if !complete
              // If we would want to poll continously, we could remove the else {}
              window.setTimeout(
                () =>
                  dispatch(
                    fetchDepartures(
                      origin,
                      destination,
                      outboundDate,
                      true,
                      index + json.departures.length
                    )
                  ),
                2000
              );
            }
          }
        });
    }
  };
}
