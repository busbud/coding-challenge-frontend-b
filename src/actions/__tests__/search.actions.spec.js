import * as actions from '../search.actions';
import { CLEAN_STORE } from '../../constants/actionTypes';

describe('cleanStore', () => {
    it('should return CLEAN_STORE', () => {
        expect(actions.cleanStore()).toEqual({
            type: CLEAN_STORE
        });
    })
})

describe('searchRequestSuccess', () => {
    it('should return SEARCH_REQUEST_SUCCESS and normalized data', () => {
        const data = {
            'departures': [
                {paris: 'montparnasse'},
                {bordeaux: 'pessac'}
            ],
            "locations": [
                {id: 12, name: 'paris 14'},
                {id: 22, name: 'centre'}
            ]
        }
        expect(actions.searchRequestSuccess(data)).toMatchSnapshot();
    })
})