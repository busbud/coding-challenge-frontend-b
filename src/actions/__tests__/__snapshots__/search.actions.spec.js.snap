// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchRequestSuccess should return SEARCH_REQUEST_SUCCESS and normalized data 1`] = `
Object {
  "departures": Array [
    Object {
      "paris": "montparnasse",
    },
    Object {
      "bordeaux": "pessac",
    },
  ],
  "locations": Object {
    "12": Object {
      "id": 12,
      "name": "paris 14",
    },
    "22": Object {
      "id": 22,
      "name": "centre",
    },
  },
  "type": "SEARCH_REQUEST_SUCCESS",
}
`;
